### PR TITLE
add basic messaging event API

### PR DESCRIPTION
**Pull request by czue**
## Summary

Context:

image

I think USH would like to pull this data into a warehouse to do some extra analysis on their side, though it's not yet clear whether this is the right data.

I modeled this off the [device log API](https://github.com/dimagi/commcare-hq/pull/3562). Guessing before we make this API "public" (if we choose to do that) we should maybe clean it up (e.g. converting the enums to more readable values, and possibly joining in related models).

## Product Description

I wasn't sure how to mark this. At the moment it is likely to be used by a single project but technically the API would be available to anyone who was able to guess the URL or is reading this PR. I don't think that's a problem unless it prevents us from making changes down the line but I wouldn't think it should do that.

Unfortunately, I don't think the infrastructure exists to feature flag an API else I'd have used a flag. Though possibly it's worth building that as it seems useful.

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

New code that won't be used by anyone apart from the project.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/api/resources/v0_5.py [hunk 1]
```diff
@@ -1087,3 +1088,23 @@ def prepend_urls(self):
...
+
+
+class MessagingEventResource(HqBaseResource, ModelResource):
+
+    class Meta(object):
+        queryset = MessagingEvent.objects.all()
+        list_allowed_methods = ['get']
+        detail_allowed_methods = ['get']
+        resource_name = 'messaging-event'
+        authentication = RequirePermissionAuthentication(Permissions.edit_data)
```
Comment (snopoke) [inline corehq/apps/api/resources/v0_5.py:1100]: any reason for the `edit_data` permission?

Comment (czue) [reply corehq/apps/api/resources/v0_5.py:1100]: Nope, just copied from [this guy](https://github.com/dimagi/commcare-hq/pull/3562/files#diff-efc09ad10beb9e39276a54e59070e11ec9947779bc6a4413196f0a086e70c841R350). It wasn't totally clear to me what it should be, and figured erring on the side of more-restrictive at first made sense.

**Review by millerdev — COMMENTED** [review]
Can you add tests for this?

I didn't think hard about the performance characteristics of this API. Like will it allow users to somehow hit the database in ways that would be disruptive to other activity? Possibly by constructing queries that perform poorly, etc.

**Comment by czue** [comment]
> Can you add tests for this?

I could if it feels worth it... can you articulate how you think they would add value? I view this as mostly a configuration of tastypie, which is presumably well-tested. As an example, the Device Log API is basically the same code and does not have tests. 

> I didn't think hard about the performance characteristics of this API. Like will it allow users to somehow hit the database in ways that would be disruptive to other activity? Possibly by constructing queries that perform poorly, etc.

I didn't think too hard either, but I don't see anything that would make it behave notably different from the existing device log API. Also, all the query params included are [also filters in the messaging events report](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/reports/standard/sms.py#L660-L738) (so those queries are already being run elsewhere in HQ - actually with more joins, etc.).

File: corehq/apps/api/resources/v0_5.py (hunk 1 above)
Comment (czue) [reply corehq/apps/api/resources/v0_5.py:1100]: @snopoke Do you think I should change this to something else?

**Comment by czue** [comment]
@millerdev @snopoke would you mind clarifying what I should do to get this approved?

File: corehq/apps/api/resources/v0_5.py (hunk 1 above)
Comment (snopoke) [reply corehq/apps/api/resources/v0_5.py:1100]: Fine for now. In the long run I think anyone who can access the messaging reports should be able to access the API as well.

**Comment by millerdev** [comment]
> can you articulate how you think they would add value?

The book club has been reading "Software Engineering at Google" where we learned about and thought it would be a good idea to adopt the [Beyoncé Rule](https://www.oreilly.com/library/view/software-engineering-at/9781492082781/ch01.html#ch01fn13-marker) (If you like it, put a [test] on it). There was an AotW that covered it as well. I think the spirit of that rule implies not adding more untested code, even if there is a precedent of other similar features that are not tested.

To clarify, I'm not talking about adding tests that cover every possible feature of tastypie that can be exercised by this API, although it might add value to test specific features you want to keep in working order. I'm thinking more of a few happy-path tests to ensure that the API as configured is operational.

**Comment by snopoke** [comment]
@czue I think at least one test would be useful that validates the API payload since changes to the payload down the line could break external systems. I am also just curious what fields are in the payload and how they are formatted.

**Comment by millerdev** [comment]
Also, the tests I'm thinking of could ideally be written in a way that was agnostic to tastypie, meaning they would not need to know anything about the internals of tastypie itself. That way if we decided to refactor or replace tastypie these tests would still be useful.

**Comment by czue** [comment]
Thanks for the clarifications. I added tests in https://github.com/dimagi/commcare-hq/pull/29308/commits/d7c189beac8f598c339a3c04600eb5247db7324a

> the spirit of that rule implies not adding more untested code, even if there is a precedent of other similar features that are not tested.

(Not related to this PR) Is this a policy that has been formally adopted? I like the philosophy but not leaving it up to discretion on a case-by-case sounds likely to lead to a lot of "silly" tests and wasted effort to me. Also, it would be good for these things to be communicated beyond the book club for those of us who don't participate (sorry if I missed an announcement).

File: corehq/apps/api/tests/test_messaging_event_api.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            "error_code": None,
+            "form_name": None,
+            "form_unique_id": None,
+            # "id": 1,  # ids are explicitly removed from comparison
+            "recipient_id": None,
+            "recipient_type": None,
+            "resource_uri": "",
+            "source": "OTH",
+            "source_id": None,
+            "status": "CMP"
```
Comment (snopoke) [inline corehq/apps/api/tests/test_messaging_event_api.py:39]: Is the meaning of these documented anywhere? Or is the plan to convert them to something more meaningful?

**Review by snopoke — APPROVED** [review]

File: corehq/apps/api/tests/test_messaging_event_api.py (hunk 2 above)
Comment (czue) [reply corehq/apps/api/tests/test_messaging_event_api.py:39]: It's all these things: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/sms/models.py#L886-L1054

Would likely figure out a way to convert them prior to "releasing" the API. initial goal is just to confirm whether this solves the client's problem.

**Comment by czue** [comment]
@millerdev going to merge since this is approved and would be nice to test soon, but happy to address any follow ups you have in a separate PR

**Review by millerdev — COMMENTED** [review]
Thanks for adding tests.

> (Not related to this PR) Is this a policy that has been formally adopted? I like the philosophy but not leaving it up to discretion on a case-by-case basis sounds likely to lead to a lot of "silly" tests and wasted effort to me. Also, it would be good for these things to be communicated beyond the book club for those of us who don't participate (sorry if I missed an announcement).

Our [code review guidelines](https://github.com/dimagi/code-review#reliable) explicitly say that "tests [should] cover new and changed code." Were you thinking of a more formal policy than that?

I think it's steadily becoming a cultural norm at Dimagi, at least on the SaaS team, that tests are expected with new code. My perception is that most of the team seems to agree that tests almost always make code better and easier to maintain. I haven't heard any complaints of "silly" tests or wasted effort, but I'd like to hear more about why you think that will happen.

**Comment by czue** [comment]
Thanks will address comments in in follow-ups.

Re: tests in general, I suspect we agree and are just using different language around the edge cases (for example, I agree adding tests for this code was a good idea and appreciate the push to do so). 

The code review guidelines also say:

```
- Asking the author to write a few simple tests for a change is a reasonable blocker.
- Asking the author to write tests that would require bootstrapping an entire test framework should not be a blocker unless the code is absolutely mission-critical (and if it is mission-critical code it likely already has a test framework).
```
Which more accurately reflects my position

Our time is a fixed resource. So while I agree with your statement:

> tests almost always make code better and easier to maintain.

I don't necessarily agree that just because that is true that it's always worth the effort, vs. e.g. fixing bugs that our customers are facing, improving the UX of our product, or cleaning up other parts of the codebase that could use more love. Each test has some amount of value and some amount of effort, and the opportunity cost is not always maximal compared to work that could be done elsewhere.

I sometimes call this the "I happened to be looking at this" fallacy. Just because there is work do be done that is good work, doesn't mean it's the *best* work.

**Comment by czue** [comment]
@snopoke guessing it was a mistake/oversight on my part. I think I attempted to figure out what a particular messaging report was backed by and ended up on `MessagingEvent` but if the requirements were to have recipient-level data then I agree based on the above that `MessagingSubEvent` seems more correct.

As you can see in the PR description this wasn't very well understood, and I had assumed that whoever was intending to use this API would provide feedback and we could then update it if there were issues. The last time I checked (which you helped with) no one was using it despite there being docs, so it may still be fine to change it. Or we could add a sub event API. Or we could add embedded sub-events to this API.
